### PR TITLE
Raft, use schema commit log

### PR DESCRIPTION
**Pull request by gusev-p**
We need this so that we can have multi-partition mutations which are applied atomically. If they live on different shards, we can't guarantee atomic write to the commitlog.

Fixes: #12642

### CODE REVIEW

**Comment by gusev-p** [comment]
@tgrabiec @kbr-scylla could you please have a look at the `schema_static_props` thing? Can't say I like this, but I don't like static `schema_ptr`-s lists either, especially that they often duplicate the configs, e.g. for `with_null_sharder`.

**Comment by kbr-scylla** [comment]
I don't think this can work for `system_distributed{_everywhere}` tables.
These tables may be created by receiving mutations from other nodes.

**Comment by gusev-p** [comment]
> so it will be initialized before the node starts booting, meaning that by the time the node receives schema definitions from other nodes, it will have populated the `local_static_props` structure, so it will apply the static properties when building schema from the received mutations. Do I understand this correctly?

Yes, but it was quite dirty. I prepared a new version, made static props more explicit.

**Comment by gusev-p** [comment]
v2:
  * move from `thread_local` dictionary with all the static properties to rule-based configs. The idea is that static props are significantly different from what is described through schema_builder. `schema_builder` specifies the schema part that should be able to turn into mutations. On the other hand, static props is what currently running node (code) thinks about how that particular table (or a set of tables) should behave.

**Comment by gusev-p** [comment]
v3:
  * removed forgotten debug log

File: db/system_keyspace.cc
Comment (kbr-scylla) [inline db/system_keyspace.cc:114]: Let's put this into a separate commit, so that the infrastructure setup and refactoring existing code to use the new infra can be reviewed / merged separately.

File: main.cc
Comment (kbr-scylla) [reply main.cc:1196]: Ah, I see you rename it later. Let's put the rename commit first or do the rename and shuffle in the same commit.

**Review by kbr-scylla — COMMENTED** [review]
I think I like the approach. @tgrabiec please review as well.

@gusev-p can you send another PR with just the new infrastructure (static schema props) + refactoring existing code, so we could merge it earlier and continue debating on the Raft commit log thing separately?

**Comment by gusev-p** [comment]
> @gusev-p can you send another PR with just the new infrastructure (static schema props) + refactoring existing code, so we could merge it earlier and continue debating on the Raft commit log thing separately?

Yes, sure.

File: main.cc
Comment (gusev-p) [reply main.cc:1196]: > This change should be explained in the commit message. 

Sure.

> And maybe the name before_schema_keyspace_init should be changed, because now it must be called even earlier, before system keyspace init, if I understand the intent?

Yes.

File: db/schema_tables.cc
Comment (tgrabiec) [inline db/schema_tables.cc:89]: Doesn't this move belong to the patch which introduces set_null_sharder?

File: db/system_keyspace.cc
Comment (tgrabiec) [reply db/system_keyspace.cc:77]: Changing things in the middle of a PR like this wastes time of reviewers, who need to figure out what changed and why. Please avoid.

File: db/system_keyspace.cc
Comment (tgrabiec) [inline db/system_keyspace.cc:94]: Why not unordered_set<> like in others of the like?

File: replica/database.cc
Comment (tgrabiec) [inline replica/database.cc:1931]: You could have kept s->wait_for_sync_to_commitlog(), just make it look at static_props(). This doesn't necessarily have to come only from static_props(), so would be more future-proof, and shorter.

File: db/schema_tables.cc
Comment (tgrabiec) [inline db/schema_tables.cc:87]: The change log should describe the change. Just "use_schema_commitlog" doesn't tell me much.

File: db/system_keyspace.cc
Comment (tgrabiec) [reply db/system_keyspace.cc:114]: Putting use_schema_commitlog in the static props and annotating raft tables with it are two separate changes.

File: main.cc
Comment (tgrabiec) [inline main.cc:1063]: This should not be moved before `db::system_keyspace::enable_features_on_startup(feature_service).get();`, because before_schema_keyspace_init() examines features.

**Review by tgrabiec — CHANGES_REQUESTED** [review]

File: main.cc
Comment (gusev-p) [reply main.cc:1063]: Yes, I already figured that out, since my new test doesn't pass.

File: replica/database.cc
Comment (gusev-p) [reply replica/database.cc:1931]: ok

File: db/system_keyspace.cc
Comment (gusev-p) [reply db/system_keyspace.cc:94]: Just moved the code "as is". Ok, will make it uniform.

File: schema/schema.cc
Comment (gusev-p) [reply schema/schema.cc:1341]: to make the syntax `const auto the_rule_name = register_static_configurator([]{ ... })` work

File: db/schema_tables.cc
Comment (tgrabiec) [reply db/schema_tables.cc:87]: The approach seems fine. You could post the static props refactoring separately, it's a nice cleanup.

**Comment by gusev-p** [comment]
@kbr-scylla @tgrabiec I created a [separate PR](https://github.com/scylladb/scylladb/pull/13170) about `schema_static_props` refactoring. It's properly structured and commented, please have a look at it.

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:704]: Maybe, I'm not sure, I've just used the pattern from `install_and_start`. There is the same pattern for `start/stop`. @alecco @kbr-scylla your are more familiar with this code, could you comment?

**Comment by gusev-p** [comment]
v5:
  * `all_system_table_load_phases`;
  * drop commitlog warning fix

File: main.cc
Comment (gleb-cloudius) [reply main.cc:1197]: I think it is reasonable to ask to shutdown node cleanly during rolling upgrade.

File: test/pylib/scylla_cluster.py
Comment (alecco) [reply test/pylib/scylla_cluster.py:704]: > Shouldn't install() cleanup after itself if it fails

But you might want to know why it failed, keeping logs.

Comment (tgrabiec) [reply test/pylib/scylla_cluster.py:704]: > But you might want to know why it failed, keeping logs.

If start() fails, then yes. But if install() fails, there's nothing in the logs yet, right?

File: main.cc
Comment (gusev-p) [reply main.cc:1197]: > I think it is reasonable to ask to shutdown node cleanly during rolling upgrade.

We need to detect somehow that the current boot is a first boot after moving raft tables to schema commitlog. As @tgrabiec mentioned above, we need to either look into the contents of commitlog while replaying it, or remember somewhere that the raft tables have already moved to a schema commitlog.

Comment (gusev-p) [reply main.cc:1197]: @tgrabiec for schema tables I also see [some logic with truncation records](https://github.com/scylladb/scylladb/blob/master/main.cc#L1236). Do we need something similar for raft tables?

Comment (gleb-cloudius) [reply main.cc:1197]: > > I think it is reasonable to ask to shutdown node cleanly during rolling upgrade.
> 
> We need to detect somehow that the current boot is a first boot after moving raft tables to schema commitlog. As @tgrabiec mentioned above, we need to either look into the contents of commitlog while replaying it, or remember somewhere that the raft tables have already moved to a schema commitlog.

I am saying that we may assume a clean shutdown during upgrade, so there cannot be a situation where non replayed raft data is in the old commitlog. Of course detecting such case would be a bonus, but I am not convinced the it is worth the complication. Currently it is not a problem even if it happens, and if it will not become a problem before 5.3 release it will never will since any upgrade will go through 5.3.

Comment (tgrabiec) [reply main.cc:1197]: > @tgrabiec for schema tables I also see [some logic with truncation records](https://github.com/scylladb/scylladb/blob/master/main.cc#L1236). Do we need something similar for raft tables?

Yes, if we want to be extra safe.

Comment (gusev-p) [reply main.cc:1197]: > Yes, if we want to be extra safe.

I think I agree with Gleb, we can describe the procedure for upgrading to 5.3 in the documentation and assume there will be no commit log replay during node boot.

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:704]: made install exception safe

**Comment by gusev-p** [comment]
v5:
  * made ScyllaServer.install exception safe

**Comment by tgrabiec** [comment]
> Wait, yet another commit log? Why can't we use the existing one?

It's about reusing the existing schema commit log. The PR title is a bit misleading.

**Comment by gusev-p** [comment]
rebased on master/CI restart

File: main.cc
Comment (gusev-p) [reply main.cc:1197]: @tgrabiec Considering Gleb’s arguments, do you agree to ask user for a clean rolling restart in the docs without enforcing this in the code? This could also make https://github.com/scylladb/scylladb/pull/13263 simpler.

In any case, it’s worth mentioning somewhere in the documentation for the 5.3 upgrade procedure that we need a clean rolling restart because of this feature. Could you advise where/how to do this? Maybe some example.

**Comment by avikivity** [comment]
> > Wait, yet another commit log? Why can't we use the existing one?
> 
> It's about reusing the existing schema commit log. The PR title is a bit misleading.

Ok. I'd like to see the cover letter explain how upgrades are handled.

**Comment by gusev-p** [comment]
> Ok. I'd like to see the cover letter explain how upgrades are handled.

This is being discussed in the comments above.

File: replica/distributed_loader.cc
Comment (kbr-scylla) [inline replica/distributed_loader.cc:729]: The commit message didn't explain this particular change. Why is it needed? What happens to the `cf`s that we skip? The code below is creating directories for them, wouldn't that mean that these `cf`s won't have directories? Why is important for them to be "ready for writes" when we create directories etc.?

A comment here would be best, but it should also be explained in the commit message.

File: main.cc
Comment (tgrabiec) [reply main.cc:1197]: We already ask for a clean shutdown in the docs for upgrades. 

Small correction: even if users shutdown cleanly, you can assume that all data is in the sstables, but not that there is no commitlog replay. The segments may be still kept alive for some reason. We shouldn't assume no replay, but I don't think this is problematic if all writes are already in the sstables. Commitlog replay doesn't annotate mutations with replay positions currently, it passes an empty `db::rp_handle`.

I'm not sure it's good to trust users in that regard. One thing is to require something, another is assuming without checking that users do something. Users are notoriously unreliable, so it's good to be defensive. @avikivity WDYT?

Comment (kbr-scylla) [reply main.cc:1197]: Are we guaranteed to flush the memtables and not replay commitlog after a clean shutdown?

IMO we should stay on the safe side if there's a reasonable not-too-hard way to do that. Don't trust the user they won't f* up.

> Could you advise where/how to do this? Maybe some example.

The documentation for 5.2->5.3 upgrade does not exist yet, we will probably create it after we release 5.2.
Once it's created, it will be similar to this one (this is for 5.1->5.2): https://github.com/scylladb/scylladb/blob/master/docs/upgrade/upgrade-opensource/upgrade-guide-from-5.1-to-5.2/upgrade-guide-from-5.1-to-5.2-generic.rst
and we'll be able to update it in a regular pull request.

We could also update the 5.1->5.2 guide to mention the necessity for clean shutdown somewhere. The 5.2->5.3 guide will most probably be created initially by copying the 5.1->5.2 guide.

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:434]: Please include the path to the log file in this message - like the message above does - to make it easier to debug a test failure.

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:442]: This interrupts the "flow" of the message, which says: "failed to start", and then gives the specifics of the failure (i.e. managed to obtain host ID and so on).

Let's put this on a new line, before "Check server log at ..." to not break the flow. Also provide more details: where was this error expected? (In the log file of the server, and not e.g. during a CQL query.)

It would be also great to include information whether the message appeared in the log. You do include it in case the process stops, but not in case of timeout which this code is handling.

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py]: As above, `None` is not a valid value for `dict[str, str]`

Comment (gusev-p) [reply test/pylib/scylla_cluster.py]: fixed

**Comment by gusev-p** [comment]
v7:
  * resolved @kbr-scylla 's review comments

File: main.cc
Comment (gusev-p) [reply main.cc:1197]: the simplest way to implement the check (IIUC):
* add a bool flag `raft_tables_uses_schema_commitlog` to some schema commitlog-managed table, e.g. as a static column in `system.raft` table with false as a default value.
* check this flag in `main.cc` after replaying schema commitlog
* if the flag is true, do nothing
* if the flag is false, check if there is something to replay in the regular commitlog (`!db.local().commitlog()->get_segments_to_replay().empty()`), fail the boot with exception if it is. User will need to rollback to previous version and ensure clean restart.
* otherwise, save true to the flag and proceed.

I'm still not sure we really need this, since right now there is no problem if we just replay the raft tables from the normal commitlog. We can add this check in a separate PR, if someone needs raft group0 before the regular commitlog replay.

Comment (gleb-cloudius) [reply main.cc:1197]: > 
> I'm still not sure we really need this, since right now there is no problem if we just replay the raft tables from the normal commitlog. We can add this check in a separate PR, if someone needs raft group0 before the regular commitlog replay.

I agree and I am do not see why the PR will be ever needed. If it is fine to replay old commitlog in 5.3 and if the dependency of regular commitlog reply on  group0 will be introduced afterwords then upgrade path is covered since it has to go through 5.3.

Comment (tgrabiec) [reply main.cc:1197]: If we require upgrade through 5.3 (with this PR), and we won't move group0 earlier in 5.3, then there is no problem.

How can we enforce staged upgrades like this? All upgrade guides of later versions will have to mention this. @mykaul  @annastuchlik

Comment (gleb-cloudius) [reply main.cc:1197]: I think we already require upgrade to not jump over minor versions.

File: test/pylib/manager_client.py
Comment (kbr-scylla) [reply test/pylib/manager_client.py]: One example of usefulness of enforcing type hints was me being sure that using `server_info.host_id` won't blow up in my face randomly at some point in CI because some code path ended up with `host_id = None`. Many tests depended on this field and assumed that it's never `None`. Now adding the possibility to only create servers but not start them means that we might sometimes not have access to the host ID. The types should reflect that so I can prepare for it when writing tests, for example. When I ask the question "can the host ID be missing?" I should be able to determine that from a single glance at the type instead of having to read the entire framework's code.

Here, in the example of our dict, I was actually hoping that you used `{}` instead of `None`, because 1. why use a larger type (`Optional[dict[str, str]]`) when you can use a smaller type (`dict[str, str]`), 2. each time we use `None` somewhere we also have to add a condition somewhere else to handle it (do this if not None else do that). If we used `dict[str, str]` the code would just merge it with the other config dict unconditionally, like it does with a non-empty config, without any `None`-`if`ology.

**Review by kbr-scylla — APPROVED** [review]
Minor stuff / questions.

File: test/pylib/manager_client.py
Comment (kbr-scylla) [reply test/pylib/manager_client.py]: > Many tests depended on this field and assumed that it's never None.

After reading the commits... that statement didn't age well.

**Comment by gusev-p** [comment]
@kbr-scylla can we queue this?

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:430]: `readlines()` also reads the entire file into memory
but now I see this is a preexisting problem, the code already uses `readlines()` in the `self.cmd.returncode` case, so let's ignore it for now

**Comment by kbr-scylla** [comment]
> @kbr-scylla can we queue this?

I'm gonna wait for @tgrabiec's approval since he requested changes earlier.
@tgrabiec please re-review

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:430]: it seems this should be fine
```
for line in open('really_big_file.dat'):
    process_data(line)
```

**Review by tgrabiec — APPROVED** [review]
Looks good. 

I only gave a cursory look at the python testing framework changes.

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:430]: ok, let's merge this as is

Comment (gusev-p) [reply test/pylib/scylla_cluster.py:430]: I'll create another PR for this

**Comment by kbr-scylla** [comment]
Queued

**Pull request merged**
